### PR TITLE
[Doc] First push of the developer documentation

### PR DESCRIPTION
**Pull request by franchuterivera**
First Push for the developer documentation.

### CODE REVIEW

**Comment by franchuterivera** [comment]
Also, add documentation for the run identifiers

File: docs/dev.rst [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+The following sections provides details regarding this two main blocks of code.
+
+Building Individual Models
+==========================
+
+AutoPyTorch relies on Scikit-Learn `Pipeline <https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html>`_ to build an individual algorithm.
+In other words, each of the individual models fitted by SMAC are (and comply) with Scikit-Learn pipeline and framework. For example, when a pipeline is fitted,
+we use pickle to save it to disk as stated `here <https://scikit-learn.org/stable/modules/model_persistence.html>`_. SMAC runs an optimization loop that proposes new
+configurations based on bayesian optimization, which comply with the package `ConfigSpace <https://automl.github.io/ConfigSpace/master/>`_. These configurations are
+translated to a pipeline configuration, fitted and saved to disc using the function evaluator `ExecuteTaFuncWithQueue`. The later is basically a worker that that
```
Comment (ravinkohli) [inline docs/dev.rst:25]: It should be the latter is basically

File: docs/dev.rst [hunk 2]
```diff
@@ -0,0 +1,53 @@
+:orphan:
+
+.. _dev:
+
+=======================
+Developer Documentation
+=======================
+
+This documentation summarizes how the AutoPyTorch code works, and it is meant to guide developers
```
Comment (ravinkohli) [inline docs/dev.rst:9]: and is meant as a guide for the developers to help contribute to it. Maybe?

File: docs/dev.rst [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+on how to best contribute to it.
+
+AutoPyTorch relies on the `SMAC <https://automl.github.io/SMAC3/master/>`_ library to build individual models,
+which are later ensembled together using ensemble selection by `Caruana et al. (2004) <https://dl.acm.org/doi/pdf/10.1145/1015330.1015432>`_.
+Therefore, there are two main parts of the code: `AutoMLSMBO`, which is our interface to the SMAC package, and
```
Comment (ravinkohli) [inline docs/dev.rst:14]: AutoMLSMBO, which acts as an interface to SMAC

File: docs/dev.rst [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+`EnsembleBuilder` which opportunistically builds an ensemble of the individual algorithms found by SMAC, at fixed intervals.
+The following sections provides details regarding this two main blocks of code.
+
+Building Individual Models
+==========================
+
+AutoPyTorch relies on Scikit-Learn `Pipeline <https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html>`_ to build an individual algorithm.
+In other words, each of the individual models fitted by SMAC are (and comply) with Scikit-Learn pipeline and framework. For example, when a pipeline is fitted,
```
Comment (ravinkohli) [inline docs/dev.rst:22]: A pipeline can consist of various preprocessing steps including- imputation, encoding, scaling, and feature preprocessing for tabular data followed by the setup of the training algorithm which includes building a network embedding for categorical features, a network backbone that is responsible for feature extraction, a network head that is responsible for decision making and choosing an appropriate Learning Rate scheduler, optimisation algorithm, and initialisation of the network parameters. 

The preprocessing steps are aggregated into a Sklearn ColumnTransformer to maintain the order of the columns and the column transformer is wrapped into a TabularColumnTransformer class which complies with a torchvision transform. Based on the size of the dataset, we apply this transformer to the whole of the dataset in the EarlyPreprocessing node if the dataset is small enough or it is added as a transform to the data loader object that allows transforming the dataset batchwise.

File: docs/dev.rst [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+we use pickle to save it to disk as stated `here <https://scikit-learn.org/stable/modules/model_persistence.html>`_. SMAC runs an optimization loop that proposes new
+configurations based on bayesian optimization, which comply with the package `ConfigSpace <https://automl.github.io/ConfigSpace/master/>`_. These configurations are
+translated to a pipeline configuration, fitted and saved to disc using the function evaluator `ExecuteTaFuncWithQueue`. The latter is basically a worker that that
```
Comment (ravinkohli) [inline docs/dev.rst:25]: These configurations are then translated to AutoPyTorch pipelines, fitted, and finally saved to disc

Comment (ravinkohli) [inline docs/dev.rst:25]: These configurations are then translated to AutoPyTorch pipelines, fitted, and finally saved to disc

File: docs/dev.rst [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+reads a dataset from disc, fits a pipeline, and collect the performance result which is communicated back to the main process via a Queue. This worker manages
+resources using `Pynisher <https://github.com/automl/pynisher>`_, and it usually does so by creating a new process.
```
Comment (ravinkohli) [inline docs/dev.rst:27]: memory and time constraints, maybe?

File: docs/dev.rst [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+The Scikit-learn pipeline inherits from the `BaseEstimator <https://scikit-learn.org/stable/modules/generated/sklearn.base.BaseEstimator.html>`_, which implies that we have to honor the `Scikit-Learn development Guidelines <https://scikit-learn.org/stable/developers/develop.html>`_. Of particular interest is that any estimator must define as attributes, the arguments that the class constructor receives (see `get_params and set_params` from the above documentation).
```
Comment (ravinkohli) [inline docs/dev.rst:29]: Particularly, the arguments that the class constructor of any estimator must be defined as attributes of the class. (see blah blah)

File: docs/dev.rst [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+Regarding multiprocessing, AutoPyTorch and SMAC work with `Dask.distributed <https://distributed.dask.org/en/latest/>`_. We only submits jobs to Dask up to the number of
```
Comment (ravinkohli) [inline docs/dev.rst:31]: To speed up the search, AutoPyTorch and SMAC use Dask.distributed which provides an excellent(or good) infrastructure to effectively multiprocess the computations involved.

File: docs/dev.rst [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+resources using `Pynisher <https://github.com/automl/pynisher>`_, and it usually does so by creating a new process.
+
+The Scikit-learn pipeline inherits from the `BaseEstimator <https://scikit-learn.org/stable/modules/generated/sklearn.base.BaseEstimator.html>`_, which implies that we have to honor the `Scikit-Learn development Guidelines <https://scikit-learn.org/stable/developers/develop.html>`_. Of particular interest is that any estimator must define as attributes, the arguments that the class constructor receives (see `get_params and set_params` from the above documentation).
+
+Regarding multiprocessing, AutoPyTorch and SMAC work with `Dask.distributed <https://distributed.dask.org/en/latest/>`_. We only submits jobs to Dask up to the number of 
+workers, and wait for a worker to be available before continuing.
+
+At the end of a SMAC runs, the results will be available in the `temporary_directory` provided to the API, in particular inside of the `<temporary_directory>/smac3-output/run_<SEED>/` directory. One can debug
```
Comment (ravinkohli) [inline docs/dev.rst:34]: run*

File: docs/dev.rst [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+the performance of the individual models using the file `runhistory.json` located in this area. Every individual model will be stored in `<temporary_directory>/.autoPyTorch/runs`. 
+In this later directory we store the fitted model (during cross-validation we store a single Voting Classifier/Regressor, which is the soft voting outcome of k-Fold cross-validation), the Out-Of-Fold
```
Comment (ravinkohli) [inline docs/dev.rst:36]: In this 'runs' directory,

File: docs/dev.rst (hunk 4 above)
Comment (ravinkohli) [reply docs/dev.rst:22]: This can also be moved to the end (AutoML part) where you talk about the pipeline

**Review by ravinkohli — CHANGES_REQUESTED** [review]
Hey, this documentation overall looks good and provides important details of the working of AutoPyTorch. I have made some comments that may improve the information we give to the developers and also minor grammar updates. All these are optional, please choose what you also see as appropriate

File: docs/dev.rst [hunk 11]
```diff
@@ -0,0 +1,54 @@
+:orphan:
+
+.. _dev:
+
+=======================
+Developer Documentation
+=======================
+
+This documentation summarizes how the AutoPyTorch code works, and is meant as a guide for the developers to help contribute to it. .
```
Comment (ravinkohli) [inline docs/dev.rst:9]: there is an extra .

**Review by ravinkohli — APPROVED** [review]
There is a minor typo but otherwise it looks great.
